File: apps/web/src/lib/roomSync.ts
import type { ChatMessage } from '../stores/sessionStore'

/**
 * 待决的幸运询价（后端 `room_sync._luck_snapshot`）。
 *
 * 为什么它需要一份快照，而战斗那几个是「HUD 显示得对不对」的问题：询价一发出，整条结算链
 * （物品发货、线索记账、KP 续写）就停在那儿等回答，`pending_luck` 是落在 turn_state 里的
 * 持久状态。可 `luck_offer` 事件是 log 类、**不落库**——玩家刷新或断线一次，那张卡就再也
 * 回不来了，人却还在等他拍板。实测有存档正是这么停在一次侦查检定上，之后一个事件都没有。
 */
export interface LuckSnapshot {
  pending: boolean
  id?: string
  char_id?: string
  actor?: string
  skill?: string
  dice_event_id?: string
  cost?: number
  available?: number
  target?: number
}

/**
 * 把快照还原成消息流里的那条幸运询价，没有待决的就返回 null。
 *
 * id 用后端给的（`luck_offer_event_id`，与检定当场那次广播同一个）——store 的
 * `addMessage` 按 id 幂等，所以在线时重连补一次也只会合成一条，不会并排出两张一样的卡。
 * 字段形状与广播时的 metadata 保持一致，两条路复用页面里同一处渲染。
 */
export function luckOfferMessage(luck: LuckSnapshot | undefined): ChatMessage | null {
  if (!luck?.pending || !luck.id || !luck.cost) return null
  return {
    id: luck.id,
    type: 'system',
    content: `${luck.actor ?? ''} 的这次检定差 ${luck.cost} 点——可花 ${luck.cost} 点幸运扭转`,
    metadata: {
      char_id: luck.char_id ?? '',
      actor: luck.actor ?? '',
      skill: luck.skill ?? '',
      dice_event_id: luck.dice_event_id ?? '',
      cost: luck.cost,
      available: luck.available,
      target: luck.target,
    },
  }
}

/**
 * `GET /sessions/{id}/sync` 的响应：sync 类状态的快照 + 事件水位线。
 *
 * 与后端 `room_sync.PROVIDERS` 一一对应——那边加一个系统，这里补一个字段。
 * `combat` 的具体形状由 CombatStage 那边的类型描述，这里只声明到调用方需要的深度，
 * 免得这个契约文件反过来依赖组件。
 */
export interface SyncSnapshot {
  seq: number
  generating: boolean
  systems: {
    combat?: { active: boolean; pending_reaction?: unknown; started_seq?: number | null }
    chase?: { active: boolean }
    turn?: { confirmed_ids: string[]; total: number; ready: boolean }
    luck?: LuckSnapshot
  }
}

File: server/app/services/room_sync.py
"""房间状态快照：``sync`` 类事件对应的真值出口。

`room_events` 把事件分成三类，其中 ``sync`` 类只承担「某状态变了」的通知，
真值在业务表里。此前这些真值散落在各自的 GET 端点上，而且并非每个系统都有：

- 战斗、追逐各有 ``GET /sessions/{id}/combat`` 与 ``/chase``；
- 回合确认（``turn_state``）**根本没有查询端点**，只能靠广播事件——断线期间错过
  那条广播，确认进度就一直是错的，直到下一次有人确认；
- 三个系统三次往返，跨网联机时每次重连要多付两个 RTT。

这里把它们收成一个注册表和一次查询。新增一个需要断线对齐的系统时，
在 ``PROVIDERS`` 里加一行即可，不必再改前端的重连流程。

``seq`` 是快照对应的事件水位线：客户端拿到快照后，可以安全地丢弃 seq 不大于它的
缓冲事件，剩下的按序应用即可对齐。
"""

from __future__ import annotations

from collections.abc import Callable

from sqlalchemy.orm import Session

from app.models.session import GameSession
from app.services import chase_service, combat_service, session_service
from app.services.event_protocol import luck_offer_event_id

# 系统名 → 取快照。键即前端 sync 事件的归属系统，两边用同一套名字。
PROVIDERS: dict[str, Callable[[Session, GameSession], dict]] = {}


def register(name: str):
    def wrap(fn: Callable[[Session, GameSession], dict]):
        PROVIDERS[name] = fn
        return fn

    return wrap


@register("combat")
def _combat_snapshot(db: Session, session: GameSession) -> dict:
    state = combat_service.get_combat(session)
    if not state:
        return {"active": False}
    return combat_service._combat_meta(state) | {"active": True}


@register("chase")
def _chase_snapshot(db: Session, session: GameSession) -> dict:
    state = chase_service.get_chase(session)
    if not state:
        return {"active": False}
    return chase_service._meta(state) | {"active": True}


@register("turn")
def _turn_snapshot(db: Session, session: GameSession) -> dict:
    # 在线集合交给调用方注入过——这里按「全体应确认者」取，重连场景下宁可多显示
    # 一个待确认者，也好过让已确认的人看不到自己的确认。
    return session_service.turn_confirm_state(db, session.id)


@register("luck")
def _luck_snapshot(db: Session, session: GameSession) -> dict:
    """待决的幸运询价。

    这一项和上面几个不一样：它不是「HUD 显示得对不对」的问题，而是**流程会不会卡死**。
    询价一旦发出，整条结算链（物品发货、线索记账、KP 续写）就停在那儿等回答，
    而 ``pending_luck`` 是落在 turn_state 里的持久状态。可 ``luck_offer`` 事件是 log 类、
    **不落库**——玩家刷新或断线一次，那张卡就再也回不来了，人却还在等他拍板。
    实测有存档正是这么停在一次侦查检定上，之后一个事件都没有。

    返回的字段与广播时的 metadata 一致，前端两条路复用同一套渲染。
    """
    pending = session_service.get_pending_luck(db, session.id) or {}
    offer = pending.get("offer") or {}
    if not pending or not offer:
        return {"pending": False}
    return {
        "pending": True,
        # 与广播同一个 id：前端按 id 幂等，重连补的与广播来的会合成一条
        "id": luck_offer_event_id(str(pending.get("dice_event_id") or "")),
        "char_id": pending.get("char_id") or "",
        "actor": pending.get("shown_name") or pending.get("disp_name") or "",
        "skill": pending.get("skill") or "",
        "dice_event_id": pending.get("dice_event_id") or "",
        "cost": offer.get("cost"),
        "reroll_cost": offer.get("reroll_cost"),
        "available": offer.get("available"),
        "target": offer.get("target"),
    }


def snapshot(db: Session, session: GameSession, systems: list[str] | None = None) -> dict:
    """取一批系统的快照。``systems`` 为空表示全取；未知系统名忽略而不是报错，
    这样前端可以先请求新系统、后端旧版本也不会 400。"""
    wanted = [s for s in (systems or PROVIDERS.keys()) if s in PROVIDERS]
    return {name: PROVIDERS[name](db, session) for name in wanted}


def current_seq(db: Session, session_id: str) -> int:
    """当前事件水位线。没有任何事件时为 0。

    直接取 max(sequence_num)，不复用 ``get_latest_events``——后者会过滤「仅 KP 可见」
    事件，若最后一条恰好是幕后推演就会返回空、把水位线错报成 0。水位线是传输层的
    序号，与可见性无关。
    """
    from sqlalchemy import func

    from app.models.event_log import EventLog

    value = (
        db.query(func.max(EventLog.sequence_num))
        .filter(EventLog.session_id == session_id)
        .scalar()
    )
    return int(value or 0)
